In [1]:
# Initial setup by importing needed modules
import pandas as pd
import numpy as np
import datetime as dt
pd.options.display.max_columns = None
pd.options.display.max_rows=None

In [2]:
# Read off data and put that into a panda dataframe
lend = pd.read_csv('data/LoanStats3a.csv', header=1, low_memory=False)

#### Understanding data structure, checking shape and data type

In [3]:
print(lend.shape)
lend.info()

(42538, 144)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42538 entries, 0 to 42537
Columns: 144 entries, id to settlement_term
dtypes: float64(115), object(29)
memory usage: 46.7+ MB


There are 144 columns, 115 of those are floats. Let's look at floats columns first. 

In [4]:
lend.select_dtypes(include='float').head()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  installment  \
0        NaN     5000.0       5000.0           4975.0       162.87   
1        NaN     2500.0       2500.0           2500.0        59.83   
2        NaN     2400.0       2400.0           2400.0        84.33   
3        NaN    10000.0      10000.0          10000.0       339.31   
4        NaN     3000.0       3000.0           3000.0        67.79   

   annual_inc  url    dti  delinq_2yrs  inq_last_6mths  \
0     24000.0  NaN  27.65          0.0             1.0   
1     30000.0  NaN   1.00          0.0             5.0   
2     12252.0  NaN   8.72          0.0             2.0   
3     49200.0  NaN  20.00          0.0             1.0   
4     80000.0  NaN  17.94          0.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       3.0      0.0   
1                     NaN                     NaN       3.0      0.0   
2                     NaN                     NaN       2.0      0.0   
3                    35.0                     NaN      10.0      0.0   
4                    38.0                     NaN      15.0      0.0   

   revol_bal  total_acc  out_prncp  out_prncp_inv   total_pymnt  \
0    13648.0        9.0        0.0            0.0   5863.155187   
1     1687.0        4.0        0.0            0.0   1014.530000   
2     2956.0       10.0        0.0            0.0   3005.666844   
3     5598.0       37.0        0.0            0.0  12231.890000   
4    27783.0       38.0        0.0            0.0   4066.908161   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16                0.00   
1          1014.53           456.46         435.17                0.00   
2          3005.67          2400.00         605.67                0.00   
3         12231.89         10000.00        2214.92               16.97   
4          4066.91          3000.00        1066.91                0.00   

   recoveries  collection_recovery_fee  last_pymnt_amnt  \
0         0.0                     0.00           171.62   
1       122.9                     1.11           119.66   
2         0.0                     0.00           649.91   
3         0.0                     0.00           357.48   
4         0.0                     0.00            67.30   

   collections_12_mths_ex_med  mths_since_last_major_derog  policy_code  \
0                         0.0                          NaN          1.0   
1                         0.0                          NaN          1.0   
2                         0.0                          NaN          1.0   
3                         0.0                          NaN          1.0   
4                         0.0                          NaN          1.0   

   annual_inc_joint  dti_joint  verification_status_joint  acc_now_delinq  \
0               NaN        NaN                        NaN             0.0   
1               NaN        NaN                        NaN             0.0   
2               NaN        NaN                        NaN             0.0   
3               NaN        NaN                        NaN             0.0   
4               NaN        NaN                        NaN             0.0   

   tot_coll_amt  tot_cur_bal  open_acc_6m  open_act_il  open_il_12m  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   

   open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  open_rv_12m  \
0          NaN                 NaN           NaN      NaN          NaN   
1          NaN                 NaN           NaN      NaN          NaN   
2          NaN                 NaN

While they are relevant information on describing the loans. We observed two issues:
1. The problem is to predict the loan default/paid off chances at the point of application. Details about the loans (such as terms, interest rates, hardship_length, etc.) are unknown upfront. Those columns cannot be used in the machine learning analysis.
2. Some columns appears to be all missing values. This bring us concerns on whether the dataset is complete. We will look into that soon.

Now we will review the 29 object columns. 

In [5]:
lend.select_dtypes(exclude='float').head()

id        term int_rate grade sub_grade                 emp_title  \
0  NaN   36 months   10.65%     B        B2                       NaN   
1  NaN   60 months   15.27%     C        C4                     Ryder   
2  NaN   36 months   15.96%     C        C5                       NaN   
3  NaN   36 months   13.49%     C        C1       AIR RESOURCES BOARD   
4  NaN   60 months   12.69%     B        B5  University Medical Group   

  emp_length home_ownership verification_status   issue_d  loan_status  \
0  10+ years           RENT            Verified  Dec-2011   Fully Paid   
1   < 1 year           RENT     Source Verified  Dec-2011  Charged Off   
2  10+ years           RENT        Not Verified  Dec-2011   Fully Paid   
3  10+ years           RENT     Source Verified  Dec-2011   Fully Paid   
4     1 year           RENT     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan                                               desc  \
0          n    Borrower added on 12/22/11 > I need to upgra...   
1          n    Borrower added on 12/22/11 > I plan to use t...   
2          n                                                NaN   
3          n    Borrower added on 12/21/11 > to pay for prop...   
4          n    Borrower added on 12/21/11 > I plan on combi...   

          purpose                 title zip_code addr_state earliest_cr_line  \
0     credit_card              Computer    860xx         AZ         Jan-1985   
1             car                  bike    309xx         GA         Apr-1999   
2  small_business  real estate business    606xx         IL         Nov-2001   
3           other              personel    917xx         CA         Feb-1996   
4           other              Personal    972xx         OR         Jan-1996   

  revol_util initial_list_status last_pymnt_d next_pymnt_d last_credit_pull_d  \
0      83.7%                   f     Jan-2015          NaN           Dec-2018   
1       9.4%                   f     Apr-2013          NaN           Oct-2016   
2      98.5%                   f     Jun-2014          NaN           Jun-2017   
3        21%                   f     Jan-2015          NaN           Apr-2016   
4      53.9%                   f     Jan-2017          NaN           Apr-2018   

  application_type hardship_flag debt_settlement_flag  \
0       Individual             N                    N   
1       Individual             N                    N   
2       Individual             N                    N   
3       Individual             N                    N   
4       Individual             N                    N   

  debt_settlement_flag_date settlement_status settlement_date  
0                       NaN               NaN             NaN  
1                       NaN               NaN             NaN  
2                       NaN               NaN             NaN  
3                       NaN               NaN             NaN  
4                       NaN               NaN             NaN

Some of the columns actually have date time or percentage values, but panda cannot recognize them in the initial load. Those columns will need data transformation.

#### Handling missing values

In [6]:
# Look at how many missing records by columns
lend.describe()

member_id     loan_amnt   funded_amnt  funded_amnt_inv   installment  \
count        0.0  42535.000000  42535.000000     42535.000000  42535.000000   
mean         NaN  11089.722581  10821.585753     10139.938785    322.623063   
std          NaN   7410.938391   7146.914675      7131.598014    208.927216   
min          NaN    500.000000    500.000000         0.000000     15.670000   
25%          NaN   5200.000000   5000.000000      4950.000000    165.520000   
50%          NaN   9700.000000   9600.000000      8500.000000    277.690000   
75%          NaN  15000.000000  15000.000000     14000.000000    428.180000   
max          NaN  35000.000000  35000.000000     35000.000000   1305.190000   

         annual_inc  url           dti   delinq_2yrs  inq_last_6mths  \
count  4.253100e+04  0.0  42535.000000  42506.000000    42506.000000   
mean   6.913656e+04  NaN     13.373043      0.152449        1.081424   
std    6.409635e+04  NaN      6.726315      0.512406        1.527455   
min    1.896000e+03  NaN      0.000000      0.000000        0.000000   
25%    4.000000e+04  NaN      8.200000      0.000000        0.000000   
50%    5.900000e+04  NaN     13.470000      0.000000        1.000000   
75%    8.250000e+04  NaN     18.680000      0.000000        2.000000   
max    6.000000e+06  NaN     29.990000     13.000000       33.000000   

       mths_since_last_delinq  mths_since_last_record      open_acc  \
count            15609.000000             3651.000000  42506.000000   
mean                35.017618               59.175568      9.343951   
std                 22.418427               47.145396      4.496274   
min                  0.000000                0.000000      1.000000   
25%                 17.000000                0.000000      6.000000   
50%                 33.000000               85.000000      9.000000   
75%                 51.000000              101.000000     12.000000   
max                120.000000              129.000000     47.000000   

            pub_rec     revol_bal     total_acc  out_prncp  out_prncp_inv  \
count  42506.000000  4.253500e+04  42506.000000    42535.0        42535.0   
mean       0.058156  1.429786e+04     22.124406        0.0            0.0   
std        0.245713  2.201844e+04     11.592811        0.0            0.0   
min        0.000000  0.000000e+00      1.000000        0.0            0.0   
25%        0.000000  3.635000e+03     13.000000        0.0            0.0   
50%        0.000000  8.821000e+03     20.000000        0.0            0.0   
75%        0.000000  1.725100e+04     29.000000        0.0            0.0   
max        5.000000  1.207359e+06     90.000000        0.0            0.0   

        total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
count  42535.000000     42535.000000     42535.000000   42535.000000   
mean   12020.388061     11312.927179      9675.675936    2239.981444   
std     9094.855348      9038.619934      7105.750304    2585.057393   
min        0.000000         0.000000         0.000000       0.000000   
25%     5465.685189      4793.115000      4400.000000     657.100000   
50%     9682.251696      8956.080000      8000.000000    1339.160000   
75%    16427.850066     15544.335000     13499.995000    2803.090000   
max    58886.473429     58563.680000     35000.020000   23886.470000   

       total_rec_late_fee    recoveries  collection_recovery_fee  \
count        42535.000000  42535.000000             42535.000000   
mean             1.516890    103.213884                14.423558   
std              7.830064    736.904244               162.335958   
min              0.000000      0.000000                 0.000000   
25%              0.000000      0.000000                 0.000000   
50%              0.000000      0.000000                 0.000000   
75%              0.000000      0.000000                 0.000000   
max            209.000000  29623.350000              7002.190000   

       last_pymnt_amnt  collections_12_mths_ex_med

The dataset has a total of 42K row of data. Some columns almost have no data at all (count = 0), and wouldn't be useful. By scanning the missing records range, I will make the decision to drop columns that have >40K missing records (in other word, records counts <2k). Loan_amnt has three missing records (which are blank rows), we will drop those as well.

In [7]:
lend_drop = lend.dropna(axis=1, how='all', thresh=40000).dropna(subset=['loan_amnt'])
lend_drop.shape

(42535, 51)

#### Why some columns have almost no data?

By looking at the columns, I come up with two reasons:
1. Sensitive information - columns such as member id are personal privacy. Lending club removed those information on purpose.
2. Information not applicable on closed loans. Column such as settlement amount may only be applicable to active loans which are still in progress on being repaid and not for loans that have been paid off (or default).

Fortunately, the two type of columns above do not appear to be relevant to the machine learning problems as they cannot be known in advance or generic information (e.g. members id) that do not have predicting power.

For the float type columns, let's have another look into what those columns are

In [8]:
lend_drop.select_dtypes(include='float').isnull().sum()

loan_amnt                        0
funded_amnt                      0
funded_amnt_inv                  0
installment                      0
annual_inc                       4
dti                              0
delinq_2yrs                     29
inq_last_6mths                  29
open_acc                        29
pub_rec                         29
revol_bal                        0
total_acc                       29
out_prncp                        0
out_prncp_inv                    0
total_pymnt                      0
total_pymnt_inv                  0
total_rec_prncp                  0
total_rec_int                    0
total_rec_late_fee               0
recoveries                       0
collection_recovery_fee          0
last_pymnt_amnt                  0
collections_12_mths_ex_med     145
policy_code                      0
acc_now_delinq                  29
chargeoff_within_12_mths       145
delinq_amnt                     29
pub_rec_bankruptcies          1365
tax_liens           

After reviewing the data dictionary for those columns, I conclude that the missing values would imply zeros. Apply fillna(0) mention to them during the data transformation steps.

#### Data Transformation

For the object columns, further data type transformations are needed. We will focus on issue_d, emp_length, earliest_cr_line, revol_util and title. Other columns are either unknown variables for machine learning or no data type conversions are needed.

#### 1. Loan issues date - convert into datetime value

In [10]:
#use assign function to avoid security type warining
lend_drop = lend_drop.assign(loan_start_d=pd.to_datetime(lend_drop.issue_d, format='%b-%Y'))

#### 2. Employment duration
- convert from text to numeric value. 
- clean up na value, assume <1 year if info is not provided (currently unemployment)

In [11]:
lend_drop['yr_emp'] = lend_drop['emp_length'].str[0:2].replace('< ',0).astype('float')

#### 3. Credit History
- Convert earliest_cr_line into from text to datetime values
- Replace missing values with zeros
- Extract year in order to obtain categorical values

In [12]:
lend_drop.earliest_cr_line = pd.to_datetime(lend_drop.earliest_cr_line, format='%b-%Y')
lend_drop['yr_credit']= (2011-lend_drop.earliest_cr_line.dt.year).fillna(0)

#### 4. Credit line % usage
- Convert to float
- Replace missing values with zeros

In [13]:
lend_drop['revol_util_dec'] = lend_drop['revol_util'].str.replace(r'%',r'0').astype('float')/100

#### 5. Title

In [14]:
# Preview on information in this column
lend_drop.title.head(15)

0                                  Computer
1                                      bike
2                      real estate business
3                                  personel
4                                  Personal
5     My wedding loan I promise to pay back
6                                      Loan
7                           Car Downpayment
8      Expand Business & Buy Debt Portfolio
9               Building my credit history.
10               High intrest Consolidation
11                            Consolidation
12                                  freedom
13                            citicard fund
14                               Other Loan
Name: title, dtype: object

Title describes the purpose of the loan and is useful information. However, in the same dataset, there is another column called 'purpose' which resembles information under Title. The purpose column has better precise label. Hence, we won't be using Title column.

In [15]:
lend_drop.purpose.value_counts()

debt_consolidation    19776
credit_card            5477
other                  4425
home_improvement       3199
major_purchase         2311
small_business         1992
car                    1615
wedding                1004
medical                 753
moving                  629
house                   426
educational             422
vacation                400
renewable_energy        106
Name: purpose, dtype: int64

#### Obtain Target variable

loan_status will serve as target for this analysis. Find out what the type of phrases exist in the loan_status column

In [16]:
lend_drop.loan_status.value_counts()

Fully Paid                                             34116
Charged Off                                             5670
Does not meet the credit policy. Status:Fully Paid      1988
Does not meet the credit policy. Status:Charged Off      761
Name: loan_status, dtype: int64

Per research through discussion forum, "Dose not meet the credit policy' are loans that had issued before Lending Club update credit policy. For the purpose of this study, we will treat them as "Fully Paid" or "Charged Off', regardless of credit policy compliance.

Convert text value feature into numerics for model training purposes. Create a dictionary to convert loan status into 0 (default) or 1 (paid off)

In [17]:
loan_dict = {'Fully Paid': 1, 'Charged Off': 0, 'Does not meet the credit policy. Status:Fully Paid': 1, 'Does not meet the credit policy. Status:Charged Off': 0 }

Create target column by mapping loan status to dictionary,

In [18]:
#use assign function to avoid security type warining
lend_drop = lend_drop.assign(target=lend_drop.loan_status.map(loan_dict))

Save cleaned data set into a csv file for quicker access.

In [19]:
lend_drop.to_csv('data/lending_clean.csv', index=False)

[Back to README](https://github.com/sittingman/lending_repayment)